### PR TITLE
SLING-8253 - Make the APIs JARs source collector able to checkout SCMs if -sources artifacts are not available

### PR DESCRIPTION
**Pull request by simonetripodi**

### CODE REVIEW

File: src/main/java/org/apache/sling/feature/maven/mojos/ApisJarMojo.java [hunk 1]
```diff
@@ -337,25 +357,92 @@ private void downloadSources(ArtifactId artifactId, File deflatedSourcesDir, Fil
...
+                        Pattern pattern = Pattern.compile((String) rewrite.getKey());
+                        Matcher matcher = pattern.matcher(connection);
+                        if (matcher.matches()) {
+                            if (matcher.groupCount() > 0) {
+                                tag = matcher.group(1);
+                            }
+
+                            connection = matcher.replaceAll((String) rewrite.getValue());
+
+                            break dance;
```
Comment (bosschaert) [inline src/main/java/org/apache/sling/feature/maven/mojos/ApisJarMojo.java]: Break dance! 💯

**Comment by bosschaert** [comment]
Hi @simonetripodi - thanks for the pull request. Since there is quite a bit of logic in the PR, I would really recommend to add some unit tests. Maybe you can isolate the logic into separate methods and write unit tests for these?

File: src/main/java/org/apache/sling/feature/maven/mojos/ApisJarMojo.java (hunk 1 above)
Comment (simonetripodi) [reply src/main/java/org/apache/sling/feature/maven/mojos/ApisJarMojo.java]: LOL this is an old joke since the College time :D

**Comment by simonetripodi** [comment]
hi there @bosschaert  - sure, just give me some time to think how to split things up, if you have a suggestion already would be much more than appreciated! TIA!

**Comment by bosschaert** [comment]
Hi @simonetripodi, as a starting point, the ApisJarMojo contains a bunch of methods: onFeature, retrieve, deflate, renameResources etc. Maybe it would be possible to already write unit tests for the methods individually. You could consider making the methods package-private to make it easier to invoke them, or otherwise you can invoke them using reflection...

**Comment by bosschaert** [comment]
Although Jenkins says that the commit cannot be built, it builds fine for me. I'm going to merge this PR.

**Comment by simonetripodi** [comment]
Thanks a lot @bosschaert for jumping in to help, very appreciated!

In order to improve my future PRs submission: how to check related Jenkins jobs? Are linked somewhere? Apologise but looks like I am not finding them! :(

**Comment by rombert** [comment]
@bosschaert @simonetripodi  - the failure was due to an error in our Jenkins/Github integration. I have re-triggered the validation job and now the commit was succesfully built.

For the record, the Jenkins job does a simple `mvn -U clean verify`. And in case you see false failures on validation please let me know, I'd like to make sure they don't happen again.

**Comment by rombert** [comment]
Also, for reference, click on the green checkmark or the red x near the commit and you will get a dialog with a "details" link

image

There you can see the full build results.

**Pull request merged**